From e637571ca767671d8114542d85bca7965e0a4251 Mon Sep 17 00:00:00 2001
From: Per Held <per.held@arm.com>
Date: Fri, 25 Apr 2025 13:25:29 +0200
Subject: [PATCH 1/2] Add got section to the DDR

---
 targets/corstone-300/platform.ld | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/targets/corstone-300/platform.ld b/targets/corstone-300/platform.ld
index d586b97..b746aa0 100644
--- a/targets/corstone-300/platform.ld
+++ b/targets/corstone-300/platform.ld
@@ -281,7 +281,7 @@ SECTIONS
 #endif
     * (expected_output_data_sec)
     * (sec_command_stream, sec_weight_data, sec_input_data)
-
+    * (.got*)
     * (ethosu_core_in_queue)
     * (ethosu_core_out_queue)
     . = ALIGN(4);
-- 
2.43.0

