From 162ea6b51bd94fabf623cc6b63cf271497eaff8d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Per=20=C3=85strand?= <per.astrand@arm.com>
Date: Fri, 13 Sep 2024 11:47:03 +0200
Subject: [PATCH] Add .data fixup from Corestone-300

---
 targets/corstone-320/platform.ld | 1 +
 1 file changed, 1 insertion(+)

diff --git a/targets/corstone-320/platform.ld b/targets/corstone-320/platform.ld
index 2010d14..fb4e7b7 100644
--- a/targets/corstone-320/platform.ld
+++ b/targets/corstone-320/platform.ld
@@ -77,6 +77,7 @@ PHDRS
     rom_boot PT_LOAD;
     rom_exec PT_LOAD;
     rom_dram PT_LOAD;
+    data     PT_LOAD; /* HACK: New prog header for .data (and friends) going in DTCM */
     null     PT_NULL;
 }
 
-- 
2.39.3 (Apple Git-146)

