From fc2ff3e005999ec185a1ae20c78c06a45651f5bc Mon Sep 17 00:00:00 2001
From: Digant Desai <digantdesai@meta.com>
Date: Mon, 2 Oct 2023 20:39:39 -0700
Subject: [PATCH 1/2] New phdr for .data section

---
 targets/corstone-300/platform.ld | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/targets/corstone-300/platform.ld b/targets/corstone-300/platform.ld
index 8d77329..8de77c4 100644
--- a/targets/corstone-300/platform.ld
+++ b/targets/corstone-300/platform.ld
@@ -94,6 +94,7 @@ PHDRS
 {
     rom_exec PT_LOAD;
     rom_dram PT_LOAD;
+    data     PT_LOAD; /* HACK: New prog header for .data (and friends) going in DTCM */
     null     PT_NULL;
 }
 
@@ -247,7 +248,7 @@ SECTIONS
     /* All data end */
     __data_end__ = .;
 
-  } > DTCM :rom_exec
+  } > DTCM :data
 
   .sram.bss :
   {
-- 
2.34.1

